### ISSUE TITLE
ftp uploading partial or empty files (depending on blocksize setting).

### ISSUE THREAD

**Issue by c0ffymachyne** [intake/filesystem_spec#173]
Hi,
Literally, I am running out of ideas here. Bug seems to be either related to flushing happening twice thus triggering empty upload overwriting previously correctly uploaded file or something else. Definitely it is an easily reproducible bug at least on my system.
Code below will upload empty or partial file. It has been literally taken from spec.AbstractFileSystem.put method:
```
    with fsspec.open('Y:/a_file.txt', "rb") as f1:
        with fsspec.open('ftp://192.168.1.160:2121/a_file.txt', "wb") as f2:
            data = True
            while data:
                data = f1.read(4096*64)
                f2.write(data)
```
Uploading file manually with ftplib seems to work fine:

```
    session = ftplib.FTP()
    session.connect('192.168.1.160',2121)
    session.login('', '')
    fp = open('Y:/a_file.txt','rb')
    session.storbinary('STOR /a_file.txt', fp, blocksize = 4096*64)
```

**Comment by martindurant** [comment]
fsspec.implementations.tests::test_write_big seems to test this sort of thing and tends to work; although the file size is not really huge. Do you have logs from the FTP server? Could it be that it doesn't respect the offset parameter in STOR? If so, then setting the block size to be bigger than the data size might help. I note that the STOR command in my implementation uses the full blocksize, the size of the buffer, but it could be much smaller.

**Comment by c0ffymachyne** [comment]
hi @martindurant , 

Can this be tested programmatically inside of fsspec's internal implementation and behavior changed accordingly?

**Comment by martindurant** [comment]
Unfortunately, the test I mentioned passes, so the server used for it is apparently more capable (or the sizes are too small to show the effect, or maybe because it's all running locally...)

**Comment by c0ffymachyne** [comment]
Tried both, removing offset from your implementation as well as full file size length, result is the same, both files uploaded are zero in size.

**Comment by martindurant** [comment]
That seems very odd, since the code really is calling STOR exactly the same as your non-fsspec working code. Could you call pdb in the `_upload_chunk` method to find out how the calls differ? Perhaps the problem is the caching of directory listings, you might want to try
```
fs = fsspec.filesystem(host="192.168.1.160", port=2121)
fs.invalidate_cache()
```
between writing and attempt to read, and see if that helps.

**Comment by c0ffymachyne** [comment]
Hi @martindurant , I think you are right. Increasing blocksize on the ftp object helped. My initial test was on readable local object but the test is actually inside of the write function and if condition is met extra flush operation is being fired which in result is making file empty (not yet sure why exactly).

```
        if self.buffer.tell() >= self.blocksize:
            self.flush()
```

That wouldn't let me go out-of-core though. How do you usually handle out-of-core data if buffer has to be bigger then the filesize ?

**Comment by martindurant** [comment]
Indeed, I had not meant increasing the block-size as a solution, merely as a way to diagnose the difference between the failure and the apparently successful direct ftplib method. 
Can you test whether your server supports the `rest` (i.e., offset) parameter on a `STOR` command?

**Comment by c0ffymachyne** [comment]
Hi, 
I should be able to look into this by the end of this year.

**Comment by martindurant** [comment]
Any progress here? The FTP callbacks are more protected now, maybe this is fixed?
